File: scala-package/core/src/main/scala/ml/dmlc/mxnet/io/MXDataIter.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ml.dmlc.mxnet.io

import ml.dmlc.mxnet.Base._
import ml.dmlc.mxnet.{DataPack, DataBatch, DataIter, NDArray, Shape}
import ml.dmlc.mxnet.IO._
import org.slf4j.LoggerFactory

import scala.collection.immutable.ListMap
import scala.collection.mutable.ListBuffer

/**
 * DataIter built in MXNet.
 * @param handle the handle to the underlying C++ Data Iterator
 */
// scalastyle:off finalize
private[mxnet] class MXDataIter(private[mxnet] val handle: DataIterHandle,
                                dataName: String = "data",
                                labelName: String = "label") extends DataIter {
  private val logger = LoggerFactory.getLogger(classOf[MXDataIter])

  // use currentBatch to implement hasNext
  // (may be this is not the best way to do this work,
  // fix me if any better way found)
  private var currentBatch: DataBatch = null

  private val (_provideData: ListMap[String, Shape],
               _provideLabel: ListMap[String, Shape],
               _batchSize: Int) =
    if (hasNext) {
      iterNext()
      val data = currentBatch.data(0)
      val label = currentBatch.label(0)
      // properties
      val res = (ListMap(dataName -> data.shape), ListMap(labelName -> label.shape), data.shape(0))
      currentBatch.dispose()
      reset()
      res
    } else {
      (null, null, 0)
    }

  private var disposed = false
  override protected def finalize(): Unit = {
    dispose()
  }

  /**
   * Release the native memory.
   * The object shall never be used after it is disposed.
   */
  def dispose(): Unit = {
    if (!disposed) {
      _LIB.mxDataIterFree(handle)
      disposed = true
    }
  }

  /**
   * reset the iterator
   */
  override def reset(): Unit = {
    currentBatch = null
    checkCall(_LIB.mxDataIterBeforeFirst(handle))
  }

  @throws(classOf[NoSuchElementException])
  override def next(): DataBatch = {
    if (currentBatch == null) {
      iterNext()
    }

    if (currentBatch != null) {
      val batch = currentBatch
      currentBatch = null
      batch
    } else {
      throw new NoSuchElementException
    }
  }

  /**
   * Iterate to next batch
   * @return whether the move is successful
   */
  private def iterNext(): Boolean = {
    val next = new RefInt
    checkCall(_LIB.mxDataIterNext(handle, next))
    currentBatch = null
    if (next.value > 0) {
      currentBatch = new DataBatch(data = getData(), label = getLabel(),
        index = getIndex(), pad = getPad())
    }
    next.value > 0
  }

  /**
   * get data of current batch
   * @return the data of current batch
   */
  override def getData(): IndexedSeq[NDArray] = {
    val out = new NDArrayHandleRef
    checkCall(_LIB.mxDataIterGetData(handle, out))
    IndexedSeq(new NDArray(out.value, writable = false))
  }

  /**
   * Get label of current batch
   * @return the label of current batch
   */
  override def getLabel(): IndexedSeq[NDArray] = {
    val out = new NDArrayHandleRef
    checkCall(_LIB.mxDataIterGetLabel(handle, out))
    IndexedSeq(new NDArray(out.value, writable = false))
  }

  /**
   * Get the index of current batch
   * @return the index of current batch
   */
  override def getIndex(): IndexedSeq[Long] = {
    val outIndex = new ListBuffer[Long]
    val outSize = new RefLong
    checkCall(_LIB.mxDataIterGetIndex(handle, outIndex, outSize))
    outIndex.toIndexedSeq
  }

  /**
   * get the number of padding examples
   * in current batch
   * @return number of padding examples in current batch
   */
  override def getPad(): MXUint = {
    val out = new MXUintRef
    checkCall(_LIB.mxDataIterGetPadNum(handle, out))
    out.value
  }

  // The name and shape of data provided by this iterator
  override def provideData: ListMap[String, Shape] = _provideData

  // The name and shape of label provided by this iterator
  override def provideLabel: ListMap[String, Shape] = _provideLabel

  override def hasNext: Boolean = {
    if (currentBatch != null) {
      true
    } else {
      iterNext()
    }
  }

  override def batchSize: Int = _batchSize
}

// scalastyle:on finalize
private[mxnet] class MXDataPack(iterName: String, params: Map[String, String]) extends DataPack {
  /**
    * get data iterator
    * @return DataIter
    */
  override def iterator: DataIter = {
    createIterator(iterName, params)
  }
}

File: csharp-package/src/MxNet/IO/MXDataIter.cs
/*****************************************************************************
   Copyright 2018 The MxNet.Sharp Authors. All Rights Reserved.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
******************************************************************************/
using System;
using MxNet.Interop;
using mx_float = System.Single;
using DataIterHandle = System.IntPtr;
using System.Linq;

// ReSharper disable once CheckNamespace
namespace MxNet.IO
{
    public sealed class MXDataIter : DataIter
    {
        private bool _debug_at_begin = true;
        private bool _debug_skip_load;
        private readonly NDArray data;
        private readonly string data_name;
        private readonly MXDataIterMap DataIterMap = new MXDataIterMap();
        private DataBatch first_batch;
        private readonly DataIterHandle handle;
        private readonly NDArray label;
        private readonly string label_name;
        private int? next_res;
        private readonly Operator op;

        public MXDataIter(string mxdataiterType, string data_name = "data", string label_name = "softmax_label")
        {
            handle = DataIterMap.GetMXDataIterCreator(mxdataiterType);
            _debug_skip_load = false;
            first_batch = Next();
            data = first_batch.Data[0];
            label = first_batch.Label[0];
            this.data_name = data_name;
            this.label_name = label_name;
            BatchSize = data.Shape[0];
            op = new Operator(handle);
        }

        public MXDataIter(DataIterHandle handle, string data_name = "data", string label_name = "softmax_label")
        {
            this.handle = handle;
            _debug_skip_load = false;
            first_batch = Next();
            data = first_batch.Data[0];
            label = first_batch.Label[0];

            BatchSize = data.Shape[0];
            op = new Operator(handle);
        }

        public override DataDesc[] ProvideData
        {
            get { return new[] {new DataDesc(data_name, data.Shape, data.DataType)}; }
        }

        public override DataDesc[] ProvideLabel
        {
            get { return new[] {new DataDesc(label_name, label.Shape, label.DataType)}; }
        }

        protected override void DisposeUnmanaged()
        {
            base.DisposeUnmanaged();
            if (handle != DataIterHandle.Zero)
                NativeMethods.MXDataIterFree(handle);
        }

        public void DebugSkipLoad()
        {
            _debug_skip_load = true;
            Console.WriteLine("Set debug_skip_load to be true, will simply return first batch");
        }

        public override NDArrayList GetData()
        {
            NativeMethods.MXDataIterGetData(handle, out var hdl);
            return new NDArray(hdl);
        }

        public override int[] GetIndex()
        {
            var r = NativeMethods.MXDataIterGetIndex(handle, out var outIndex, out var outSize);
            Logging.CHECK_EQ(r, 0);

            var outIndexArray = InteropHelper.ToUInt64Array(outIndex, (uint) outSize);
            var ret = new int[outSize];
            for (var i = 0ul; i < outSize; ++i)
                ret[i] = (int) outIndexArray[i];

            return ret;
        }

        public override NDArrayList GetLabel()
        {
            NativeMethods.MXDataIterGetLabel(handle, out var hdl);
            return new NDArray(hdl);
        }

        public override int GetPad()
        {
            var r = NativeMethods.MXDataIterGetPadNum(handle, out var @out);
            return @out;
        }

        public override bool IterNext()
        {
            if (first_batch != null)
                return true;

            next_res = 0;
            NativeMethods.MXDataIterNext(handle, out next_res);
            return Convert.ToBoolean(next_res.Value);
        }

        public override bool End()
        {
            throw new NotImplementedException();
        }

        public override void Reset()
        {
            _debug_at_begin = true;
            first_batch = null;
            NativeMethods.MXDataIterBeforeFirst(handle);
        }

        public override DataBatch Next()
        {
            if (_debug_skip_load && !_debug_at_begin) return new DataBatch(GetData(), GetLabel(), GetPad(), GetIndex());

            if (first_batch != null)
            {
                var batch = first_batch;
                first_batch = null;
                return batch;
            }

            _debug_at_begin = false;
            next_res = 0;
            NativeMethods.MXDataIterNext(handle, out next_res);
            if (next_res.HasValue)
                return new DataBatch(GetData(), GetLabel(), GetPad(), GetIndex());
            throw new MXNetException("Stop Iteration");
        }

        public void SetParam(string key, string value)
        {
            op.SetParam(key, value);
        }

        public void SetParam(string key, NDArray value)
        {
            op.SetParam(key, value);
        }

        public void SetParam(string key, Symbol value)
        {
            op.SetParam(key, value);
        }

        public void SetParam(string key, object value)
        {
            op.SetParam(key, value);
        }

        public NDArrayList GetItems()
        {
            var output_vars = new NDArrayList();
            NativeMethods.MXDataIterGetItems(this.handle, out var num_output, output_vars.Handles);
            return output_vars;
        }

        public int Length
        {
            get
            {
                NativeMethods.MXDataIterGetLenHint(this.handle, out var length);
                if (length < 0)
                {
                    return 0;
                }

                return length;
            }
        }
    }
}